betty/plugin/__init__.py

```diff
@@ -0,0 +1,152 @@
+"""
+The Plugin API.
+
+Plugins allow third-party code (e.g. your own Python package) to add functionality
+to Betty.
+
+Read more at :doc:`/usage/plugin`.
+"""
+
+from __future__ import annotations
+
+from typing import TypeVar, Generic, Self, overload, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from betty.locale import Localizable
+    from collections.abc import AsyncIterator, Sequence
+
+
+class Plugin:
+    """
+    A plugin.
+    """
+
+    @classmethod
+    def plugin_id(cls) -> str:
+        """
+        Get the plugin ID.
+
+        IDs are unique per plugin type:
+        - A plugin repository MUST at most have a single plugin for any ID.
+        - Different plugin repositories MAY each have a plugin with the same ID.
+        """
+        raise NotImplementedError(repr(cls))
+
+    @classmethod
+    def plugin_label(cls) -> Localizable:
+        """
+        Get the human-readable short plugin label.
+        """
+        raise NotImplementedError(repr(cls))
+
+    @classmethod
+    def plugin_description(cls) -> Localizable | None:
+        """
+        Get the human-readable long plugin description.
+        """
+        return None
+
+
+_PluginT = TypeVar("_PluginT", bound=Plugin)
+
+
+class PluginNotFound(ValueError):
+    """
+    Raised when a plugin cannot be found.
+    """
+
+    @classmethod
+    def new(cls, plugin_id: str) -> Self:
+        """
+        Create a new instance.
+        """
+        return cls(f'Could not find a plugin "{plugin_id}".')
+
+
+_PluginMixinOneT = TypeVar("_PluginMixinOneT")
+_PluginMixinTwoT = TypeVar("_PluginMixinTwoT")
+_PluginMixinThreeT = TypeVar("_PluginMixinThreeT")
+
+
+class PluginRepository(Generic[_PluginT]):
+    """
+    Discover and manage plugins.
+    """
+
+    async def get(self, plugin_id: str) -> type[_PluginT]:
+        """
+        Get a single plugin by its ID.
+
+        :raises PluginNotFound: if no plugin can be found for the given ID.
+        """
+        raise NotImplementedError(repr(self))
+
+    @overload
+    async def select(self) -> Sequence[type[_PluginT]]:
+        pass  # pragma: no cover
+
+    @overload
+    async def select(
+        self, mixin_one: type[_PluginMixinOneT]
+    ) -> Sequence[type[_PluginT & _PluginMixinOneT]]:
+        pass  # pragma: no cover
+
+    @overload
+    async def select(
+        self, mixin_one: type[_PluginMixinOneT], mixin_two: type[_PluginMixinTwoT]
+    ) -> Sequence[type[_PluginT & _PluginMixinOneT & _PluginMixinTwoT]]:
+        pass  # pragma: no cover
+
+    @overload
+    async def select(
+        self,
+        mixin_one: type[_PluginMixinOneT],
+        mixin_two: type[_PluginMixinTwoT],
+        mixin_three: type[_PluginMixinThreeT],
+    ) -> Sequence[
+        type[_PluginT & _PluginMixinOneT & _PluginMixinTwoT & _PluginMixinThreeT]
+    ]:
+        pass  # pragma: no cover
+
+    async def select(
```

Comment:
Document the reason for these arguments: we cannot unpack variable arguments into an intersection